src/btree/bt_sync.c

```diff
@@ -194,11 +234,12 @@ __sync_ref_evict_or_mark_deleted(WT_SESSION_IMPL *session, WT_REF *ref)
         return (__wt_page_parent_modify_set(session, ref, true));
     }
 
-    /* Evict the in-memory obsolete page */
-    WT_IGNORE_RET_BOOL(__wt_page_evict_urgent(session, ref));
-    WT_STAT_CONN_INCR(session, hs_gc_pages_evict);
+    /* Store the in-memory obsolete page */
+    dup = NULL;
+    WT_RET(__sync_dup_walk(session, ref, flags, &dup));
+    WT_RET(__sync_ref_list_add(session, rlp, dup));
```

Comment:
Because `__sync_dup_walk()` can return `dup == NULL` and a return status of 0, this feels dangerous to me, can't we potentially add a `NULL` to the list of pages to eventually be evicted?

(I didn't really try and figure it out -- I'd rather just not have to worry about it.)

Reply:
`__sync_dup_walk()` can set `dup == NULL` only when the `ref == NULL` or `ref` is root. we never call the `__sync_dup_walk()` function in that flow. As we never mark the root page as obsolete. So it shouldn't be a problem.

Reply:
Yeah, but what happens when somebody changes `__sync_dup_walk()` in the future? It's fragile.

Reply:
I added a new function just to get the hazard pointer to avoid future change problems.